File: LeetCode Problems/201 to 300/208. Implement Trie (Prefix Tree)/208. Implement Trie (Prefix Tree).cpp
class Trie {
public:
    class Node {
    public:
        char ch;
        bool isWord;
        Node *children[26];
        Node(char ch) {
            this->ch = ch;
            isWord = false;
            for (int i = 0; i < 26; i++)
                children[i] = nullptr;
        }
    };
    Node *root;
    Trie() { root = new Node('-'); }

    void insert(string word) {
        Node *node = root;
        for (int i = 0; i < word.size(); i++) {
            if (node->children[word[i] - 'a'] == nullptr) {
                node->children[word[i] - 'a'] = new Node(word[i]);
            }
            node = node->children[word[i] - 'a'];
        }
        node->isWord = true;
    }
    Node *getNode(string word) {
        Node *node = root;
        for (int i = 0; i < word.size(); i++) {
            if (node->children[word[i] - 'a'] == nullptr) {
                return nullptr;
            }
            node = node->children[word[i] - 'a'];
        }
        return node;
    }

    bool search(string word) {
        Node *node = getNode(word);
        return node != nullptr && node->isWord;
    }

    bool startsWith(string prefix) { return getNode(prefix) != nullptr; }
};

File: LeetCode Problems/201 to 300/208. Implement Trie (Prefix Tree)/208. Implement Trie (Prefix Tree).java
class Trie {
    class Node {
        public char ch;
        public boolean isWord;
        public Node children[];
        public Node(char ch) {
            this.ch = ch;
            isWord = false;
            children = new Node[26];
        }
    }
    private Node root;

    public Trie() {
        root = new Node('-');
    }

    public void insert(String word) {
        Node node = root;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            if (node.children[ch - 'a'] == null) {
                node.children[ch - 'a'] = new Node(ch);
            }
            node = node.children[ch - 'a'];
        }
        node.isWord = true;
    }
    private Node getWord(String word) {
        Node node = root;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            if (node.children[ch - 'a'] == null) {
                return null;
            }
            node = node.children[ch - 'a'];
        }
        return node;
    }

    public boolean search(String word) {
        Node node = getWord(word);
        return node != null && node.isWord;
    }

    public boolean startsWith(String prefix) {

        return getWord(prefix) != null;
    }
}
